controllers/storagecluster/provider_server.go

```diff
@@ -0,0 +1,272 @@
+		}
+	}
+
+	if finalErr == nil {
+		r.Log.Info("nigoyal Resource deletion for provider succeeded")
+	}
+
+	return finalErr
+}
+
+func (o *ocsProviderServer) createDeployment(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredDeployment := GetProviderAPIServerDeployment(instance)
+	actualDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      desiredDeployment.Name,
+			Namespace: desiredDeployment.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualDeployment,
+		func() error {
+			actualDeployment.Spec = desiredDeployment.Spec
+			return controllerutil.SetOwnerReference(instance, actualDeployment, r.Client.Scheme())
+		},
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.Log.Error(err, "nigoyal Failed to create/update deployment", "Name", desiredDeployment.Name)
+		return err
+	}
+
+	err = o.isDeploymentReady(actualDeployment, desiredDeployment)
+	if err != nil {
+		r.Log.Error(err, "nigoyal Deployment is not ready", "Name", desiredDeployment.Name)
+		return err
+	}
+
+	r.Log.Info("nigoyal Deployment is running as desired")
+	return nil
+}
+
+func (o *ocsProviderServer) createService(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredService := GetProviderAPIServerService(instance)
+	actualService := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      desiredService.Name,
+			Namespace: desiredService.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(
+		context.TODO(), r.Client, actualService,
+		func() error {
+			actualService.Spec = desiredService.Spec
+			return controllerutil.SetOwnerReference(instance, actualService, r.Client.Scheme())
+		},
+	)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		r.Log.Error(err, "nigoyal Failed to create/update service", "Name", desiredService.Name)
+		return err
+	}
+
+	r.Log.Info("nigoyal Service create/update succeeded")
+	return nil
+}
+
+func (o *ocsProviderServer) createSecret(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+
+	desiredSecret := GetProviderAPIServerSecret(instance)
+	actualSecret := &corev1.Secret{}
+
+	err := r.Client.Get(context.TODO(), client.ObjectKeyFromObject(desiredSecret), actualSecret)
+
+	if err != nil && errors.IsNotFound(err) {
+		err = r.Client.Create(context.TODO(), desiredSecret)
+		if err != nil {
+			r.Log.Error(err, "nigoyal Failed to create Secret", "Name", desiredSecret.Name)
+			return err
+		}
+		r.Log.Info("nigoyal Secret create succeeded", "Name", desiredSecret.Name)
+	} else if err != nil {
+		r.Log.Error(err, "nigoyal Failed to get Secret", "Name", desiredSecret.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (o *ocsProviderServer) isDeploymentReady(actual, desired *appsv1.Deployment) error {
+
+	if actual.Status.AvailableReplicas != *desired.Spec.Replicas {
+		return fmt.Errorf("Deployment %s is not ready", desired.Name)
+	}
```

Comment:
I also like this! Look how simple it is. :grin: For a future PR, I would like to see the following:

* Rather than just returning `error`, I would suggest `bool, error`. Yes checking for `err != nil` achieves more or less the same thing, but for most any function starting with `is*` it makes semantic sense for it to be treated as or return a boolean. 
* In the error message, include what criteria weren't met for the Deployment to be considered "ready".

Reply:
Why `bool, error` for a sync function that works on memory only. I think `is` function should only return `bool` and the log or error should be created from the outside context. That way the util is more usable for other use cases where you don't actually need an error. For example, think about the `!is` use case

Reply:
> For example, think about the `!is` use case

Exactly this! Specifically, though, if the `is` function grows to more than just a single check (which it is now), we should understand why the function returned false when it does. And yes, we could effectively log the error in-place and return false, but then what does the human debugging / RCA experience look like? And error message gets propagated back up and could be exposed in the CR Status, but a log message is just a log. I think?

Reply:
> if the is function grows to more than just a single check 
So it is not an `is` function anymore, if you insist you can also return a enum string for a reason. Again, no need for error

Reply:
I changed the name of the func itself to the `ensureDeploymentReplica` than the func structure.